### DISCUSSION TITLE
Edge/Chrome Browser Extension to Auto Add Recipe

### DISCUSSION THREAD

**Discussion by KevinKTM** [mealie-recipes/mealie#4457] [Feature Request]
### First Check

- [X] I used the GitHub search to find a similar requests and didn't find it.
- [X] Checked the [tasks tagged](https://github.com/mealie-recipes/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

### Please provide a concise description of the problem that would be addressed by this feature.

The process of creating a recipe involves copying the url then opening up mealie, logging in and clicking the add button. This process could be expedited with a browser addition. 

### Please provide a concise description of the feature that would resolve your issue.

The browser integration would take the URL of the current recipe you are viewing, open up a mealie instance and create the recipe. The extension would need a config page to setup the path to your mealie instance, username and password to allow automated login. Would probably be nice to have options to just add the recipe without any interaction and then also have an interactive mode that allows the user to check the recipe is scraped properly. I usually specify stay in edit mode and all recipe imports so that I can setup categories and tags. This would significantly expedite the recipe creation process. 

### Please consider and list out some caveats or tradeoffs made in your design decision

none

### Additional Information

- [ ] If this is accepted I'm willing to submit a PR to provide this feature
- [ ] If this is accepted I'm willing to help maintain this feature
- [ ] I'm willing to sponsor/pay a developer to do this work

**Comment by Kuchenpirat** [comment]
You are able to add an bookmarklet which is close to what you are asking. 
See: https://docs.mealie.io/documentation/community-guide/import-recipe-bookmarklet/

But an integration via a browser extention would be also pretty nice 👍

**Comment by burgerga** [comment]
I was about to suggest this feature. It has one additional benefit: you can use the html that is loaded in your browser to create the recipe. Why is this useful? Some websites have scraping protection (e.g. TLS fingerprinting, as used by ah.nl, see https://github.com/mealie-recipes/mealie/issues/2888), and then "Import from URL" doesn't work. Using a browser extension you could then use the raw HTML/JSON to create the recipe anyway, avoiding the not-so-user-friendly steps I wrote down in https://github.com/mealie-recipes/mealie/issues/2888#issuecomment-2558565783.

As an added bonus, the Mealie extension could then easily provide the same functionality as https://chromewebstore.google.com/detail/recipe-filter/ahlcdjbkdaegmljnnncfnhiioiadakae?hl=en (skipping the blabla, and displaying the recipe straight away).

(this is the same problem that all the bookmark managers (e.g., linkwarden, readeck, hoarder) face; articles behind a paywall can't be saved. Only readeck has solved it with a [Readeck browser extension](https://chromewebstore.google.com/detail/readeck/jnmcpmfimecibicbojhopfkcbmkafhee), which is how I got inspired about a Mealie extension :) )

**Reply by pathia** [reply]
Meanwhile we could also think of a tampermonkey script which grabs the schema.org recipe code and copies it to the clipboard.

**Reply by pathia** [reply]
I'm not that experienced with tampermonkey (greasemonkey) but I've made this as a proof of concept:
```
// ==UserScript==
// @name         AH Allerhande recept JSON
// @namespace    http://tampermonkey.net/
// @version      2025-02-04
// @description  copy the raw recipe to your clipboard
// @author       pathia
// @match        https://www.ah.nl/allerhande/recept/*
// @icon         https://www.google.com/s2/favicons?sz=64&domain=ah.nl
// @grant        window.onurlchange
// @grant        GM.setClipboard
// ==/UserScript==

(function() {
    'use strict';
    const h1 = document.querySelector('h1')

    var copyButton = document.createElement('button');
    copyButton.textContent = 'COPY JSON';
    copyButton.addEventListener('click', function(e) {
        e.preventDefault();
        const jsoncode = document.querySelectorAll('script[type="application/ld+json"]')[3];
         GM.setClipboard(jsoncode.textContent)
        h1.textContent = 'json copied !\n' + h1.textContent
    });

    h1.appendChild(copyButton);

    })();
```

**Reply by smarthomejunkie** [reply]
> I'm not that experienced with tampermonkey (greasemonkey) but I've made this as a proof of concept:
> 
> ```
> // ==UserScript==
> // @name         AH Allerhande recept JSON
> // @namespace    http://tampermonkey.net/
> // @version      2025-02-04
> // @description  copy the raw recipe to your clipboard
> // @author       pathia
> // @match        https://www.ah.nl/allerhande/recept/*
> // @icon         https://www.google.com/s2/favicons?sz=64&domain=ah.nl
> // @grant        window.onurlchange
> // @grant        GM.setClipboard
> // ==/UserScript==
> 
> (function() {
>     'use strict';
>     const h1 = document.querySelector('h1')
> 
>     var copyButton = document.createElement('button');
>     copyButton.textContent = 'COPY JSON';
>     copyButton.addEventListener('click', function(e) {
>         e.preventDefault();
>         const jsoncode = document.querySelectorAll('script[type="application/ld+json"]')[3];
>          GM.setClipboard(jsoncode.textContent)
>         h1.textContent = 'json copied !\n' + h1.textContent
>     });
> 
>     h1.appendChild(copyButton);
> 
>     })();
> ```

This works fine for me.

**Reply by burgerga** [reply]
Awesome!

Made some small improvements (also not so familiar with tampermonkey, so I used chatGPT :P):
* Not hardcode the Recipe script index
* Remove the "stap 1,2,3" names from the instructions, otherwise you get those as separate steps in Mealie

```
(function() {
    'use strict';
    const h1 = document.querySelector('h1');

    const copyButton = document.createElement('button');
    copyButton.textContent = 'COPY JSON';
    copyButton.addEventListener('click', function(e) {
        e.preventDefault();

        const scripts = document.querySelectorAll('script[type="application/ld+json"]');
        let recipe = null;

        for (const script of scripts) {
            try {
                const data = JSON.parse(script.textContent);
                if (data['@context'] === 'https://schema.org' && data['@type'] === 'Recipe') {
                    recipe = data;
                    break;
                }
            } catch (err) {
                continue;
            }
        }

        if (!recipe) {
            h1.textContent = 'No Recipe JSON found\n' + h1.textContent;
            return;
        }

        // delete step names to prevent useless extra steps in Mealie
        if (Array.isArray(recipe.recipeInstructions)) {
            recipe.recipeInstructions.forEach(step => {
                if (step.name && step.name.toLowerCase().startsWith('stap')) {
                    delete step.name;
                }
            });
        }

        const result = JSON.stringify(recipe, null, 2); //prettify result
        GM.setClipboard(result);

        h1.textContent = 'Cleaned Recipe JSON copied!\n' + h1.textContent;
    });

    h1.appendChild(copyButton);
})();
```

**Reply by mrshappy0** [reply]
@KevinKTM I haven't promoted it but I already started on a [chrome extension](https://chromewebstore.google.com/detail/mini-mealie/lchfnbjpjoeejalacnpjnafenacmdocc) mostly for fun. I've made it [open source](https://github.com/mrshappy0/mini-mealie) so you can snag the code and modify it how you please. I'm open to working with Mealie someday and adding a PR to add it to the Mealie eco-system. I built it with WXT and it only currently builds for the Chrome Web Store but I aim to add additional browsers to the list. 

It's only been live for a few months and features include the following:
- HTTPS/HTTP toggle to accommodate local or public Mealie instances
- Utilizes the get user endpoint to verify server connection within extension popup
- Paywall ladder toggle. I learned shortly after creating this that mealie does a good job of jumping NYT recipe paywall. So this one is a bit redundant for now.

In-progress: 
- Working on using the test create recipe endpoint to intelligently add "create a recipe" to the users chrome menu based on the dry-run of "create recipe"
- Some kind of log page for "create a recipe" attempts
- Ability to delete a successfully created recipe

For now, I don't have a web extension landing page for feedback but if you add the chrome extension, you can add feedback on the chrome web store or in the GitHub project. Also I made it open-source so people can happily clone/fork and/or contribute. Still working on contribution docs.

I'm not trying to steal Mealie's thunder. I have deep respect for this project and just wanted to extend it. If the project maintainers feel it's out of line, I'd be happy to chat to work on adding it directly to the Mealie repositories. I landed on this discussion because I was trying to understand how I missed the paywall abilities of Mealie.

**Reply by burgerga** [reply]
Hi @mrshappy0, that is a nice start! Maybe you can add your comment as a separate thread? Because I see from the console that you use the `/api/recipes/create/url` endpoint, which is precisely the endpoint that would not work for the issue discussed in this comment thread (this thread is about creating the recipe by extracting the schema.org/Recipe json).

About your extension: do you have some guidance on how to best sort out CORS issues? I'm succesfully connected, but API calls are blocked...

**Reply by mrshappy0** [reply]
@burgerga I will get back you on this. But I use authelia and traefik. It was a CORS configuration change to allow the extension to hit my public instance of mealie. Did you end up resolving that? Are you serving your Mealie instance publically? I can try to dig my configuration related to that.
